### PR TITLE
Add a role for RoboCup 2021 with appropriate logging

### PR DESCRIPTION
**Pull request by JosephusPaye**

### CODE REVIEW

File: docker/nugus_sim/run.py [hunk 1]
```diff
@@ -100,6 +101,25 @@ def update_config_files(args: dict) -> None:
...
+        # Ensure the log directory exists
+        Path(robot_log_dir).mkdir(parents=True, exist_ok=True)
+
+        # Configure FileLogHandler to write to the RoboCup log directory
+        file_log_handler_config = read_config("FileLogHandler.yaml")
+        file_log_handler_config["log_file"] = f"{robot_log_dir}/log"
+        write_config("FileLogHandler.yaml", file_log_handler_config)
+
+        # Configure DataLogging to write to the RoboCup log directory
+        data_logging_config = read_config("DataLogging.yaml")
```
Comment (JosephusPaye) [inline docker/nugus_sim/run.py:119]: Should this be reading and writing `robocup2021/DataLogging.yaml` (i.e. the role-specific config)?

File: roles/robocup2021.role [hunk 2]
```diff
@@ -0,0 +1,45 @@
+nuclear_role(
```
Comment (JosephusPaye) [inline roles/robocup2021.role:1]: Let me know if I've missed anything that should or shouldn't be in here.

**Comment by JosephusPaye** [comment]
Looks like DataLogging is failing to compile due to the issue fixed in #473.

**Comment by Y0sh1dk** [comment]
I think @ysims suggested earlier that the role be named something like `webotsrobocup` instead of something specific to this year, particularly if we end up using webots again next year.

File: roles/robocup2021.role [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  # Support
+  support::configuration::GlobalConfig
+  network::NUClearNet
+  # support::NUsight
+  support::configuration::SoccerConfig
+  # input
+  input::GameController
+  # Simulator connection
+  platform::Webots
+  input::SensorFilter
```
Comment (ysims) [inline roles/robocup2021.role:18]: For some reason webots needs to be after sensorfilter, or it wont work. Was probably the issue in the lab today.

File: docker/nugus_sim/run.py (hunk 1 above)
Comment (Bidski) [reply docker/nugus_sim/run.py:119]: It can, but doesn't need to be

File: docker/nugus_sim/run.py [hunk 4]
```diff
@@ -4,6 +4,7 @@
 import os
 import subprocess
 import sys
+from pathlib import Path
```
Comment (Bidski) [inline docker/nugus_sim/run.py:7]:
```suggestion
```

File: docker/nugus_sim/run.py [hunk 5]
```diff
@@ -100,6 +101,25 @@ def update_config_files(args: dict) -> None:
...
     write_config("webots.yaml", webots_config)
 
+    # Configure logging to /robocup-logs if it exists
+    if os.path.exists("/robocup-logs"):
+        # Compute the log directory for this robot, based on ROBOCUP_ROBOT_ID
+        hostname = f"webots{env_vars['ROBOCUP_ROBOT_ID']}"
+        robot_log_dir = f"/robocup-logs/{hostname}"
+
+        # Ensure the log directory exists
+        Path(robot_log_dir).mkdir(parents=True, exist_ok=True)
```
Comment (Bidski) [inline docker/nugus_sim/run.py:111]:
```suggestion
        os.makedirs(robot_log_dir, exist_ok=True)
```

File: docker/nugus_sim/run.py [hunk 6]
```diff
@@ -100,6 +101,25 @@ def update_config_files(args: dict) -> None:
     webots_config["port"] = int(port)
     write_config("webots.yaml", webots_config)
 
+    # Configure logging to /robocup-logs if it exists
+    if os.path.exists("/robocup-logs"):
```
Comment (Bidski) [inline docker/nugus_sim/run.py:105]: Why check if it exists? Is this meant to be a mounted volume?

File: module/support/logging/DataLogging/data/config/robocup2021/DataLogging.yaml [hunk 7]
```diff
@@ -0,0 +1,10 @@
+output:
+  directory: log
+  split_size: 5 * GiB
+
+# You can put any messages that is emitted in the system here and it will be logged
+# The below ones are just the most common and here as an example
+messages:
+  message.output.CompressedImage: true
```
[the remark this answers is not in the document]
Comment (Bidski) [reply module/support/logging/DataLogging/data/config/robocup2021/DataLogging.yaml:8]: - Vision balls and goals (so we can check to see that we are getting detections)
- GreenHorizon would be nice, but is also a large message (visual mesh too, but also a large message)
- Localisation messages

File: roles/webotsrobocup.role [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+  # FileWatcher, FileLogHandler and Signal Catcher Must Go First. KinematicsConfiguration usually goes after these and
+  # without it many roles do not run
+  extension::FileWatcher
+  support::SignalCatcher
+  support::logging::FileLogHandler
+  # This must come first as it emits config which many roles depend on (e.g. SensorFilter, WalkEngine)
+  motion::KinematicsConfiguration
+  # Support
+  support::configuration::GlobalConfig
+  network::NUClearNet
```
Comment (Bidski) [inline roles/webotsrobocup.role:11]: Not sure we need NUClearNet? Unless we plan on trying to organise inter-robot communication? @TrentHouliston

File: roles/webotsrobocup.role [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+  # Behaviour
+  behaviour::Controller
+  behaviour::skills::Stand
+  behaviour::skills::FallingRelax
+  behaviour::skills::Getup
+  behaviour::skills::KickScript
+  behaviour::skills::HeadBehaviourSoccer
+  behaviour::planning::KickPlanner
+  behaviour::planning::SimpleWalkPathPlanner # Creates WalkPaths. behaviour::planning::BezierWalkPathPlanner  # Creates
+                                             # WalkPaths.
```
Comment (Bidski) [inline roles/webotsrobocup.role:36]:
```suggestion
  behaviour::planning::SimpleWalkPathPlanner # Creates WalkPaths.
```

File: roles/webotsrobocup.role [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+  behaviour::strategy::SoccerStrategy
+  behaviour::skills::Nod
```
Comment (Bidski) [inline roles/webotsrobocup.role:38]: Probably not needed

**Review by Bidski — CHANGES_REQUESTED** [review]

File: docker/nugus_sim/run.py (hunk 6 above)
Comment (JosephusPaye) [reply docker/nugus_sim/run.py:105]: Yep, it will be a mounted volume which may or may not be there

File: docker/nugus_sim/run.py (hunk 1 above)
Comment (JosephusPaye) [reply docker/nugus_sim/run.py:119]: OK

File: module/support/logging/DataLogging/data/config/webotsrobocup/DataLogging.yaml [hunk 11]
```diff
@@ -0,0 +1,17 @@
+output:
+  directory: log
+  split_size: 5 * GiB
+
+# You can put any messages that is emitted in the system here and it will be logged
+messages:
+  message.output.CompressedImage: true
```
Comment (JosephusPaye) [inline module/support/logging/DataLogging/data/config/webotsrobocup/DataLogging.yaml:7]: As I understand we'll be getting `Image`s from webots, not `CompressedImage`s. So we probably need to setup the image compressor to do the conversion, if it isn't already.

File: module/support/logging/DataLogging/data/config/robocup2021/DataLogging.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (JosephusPaye) [reply module/support/logging/DataLogging/data/config/robocup2021/DataLogging.yaml]: Added all of those. After we get the docker image working with webots we should run the robocup role and see how much log space we use with those large messages enabled.

File: module/support/logging/DataLogging/data/config/webotsrobocup/DataLogging.yaml (hunk 11 above)
Comment (ysims) [reply module/support/logging/DataLogging/data/config/webotsrobocup/DataLogging.yaml:7]: I think we just need to have ImageCompressor in the role and it'll do the compression when it gets an image? Then it'll log that image.

File: roles/webotsrobocup.role [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+  vision::GreenHorizonDetector
+  vision::GoalDetector
+  vision::BallDetector
+  # Localisation
+  localisation::BallLocalisation
+  localisation::RobotParticleLocalisation
+  # Behaviour
+  behaviour::Controller
+  behaviour::skills::Stand
+  behaviour::skills::FallingRelax
```
Comment (ysims) [inline roles/webotsrobocup.role:30]: We shouldn't have falling relax - the robot wont break itself anyway, and relaxing the joints currently just sends NaN to webots which is not happy at getting NaNs!

File: module/support/logging/DataLogging/data/config/webotsrobocup/DataLogging.yaml (hunk 11 above)
Comment (JosephusPaye) [reply module/support/logging/DataLogging/data/config/webotsrobocup/DataLogging.yaml:7]: Yep, seems so.

File: roles/webotsrobocup.role (hunk 8 above)
Comment (JosephusPaye) [reply roles/webotsrobocup.role:11]: Commenting out for now.

**Review by Bidski — APPROVED** [review]
Sure

**Pull request merged**